Remove user from docker run command

antoniobeyah opened the pull request:
Fixes #25 , removes the user from the docker run command.  The changes from #25  were merged in and tested against the demo container.  I also updated the Dockerfile in the demo in preparation of the release to support the new required minimum version (Docker client 1.7.0). During testing, I found a bug in master with docker client 1.7.0, the bug was introduced in [this commit](https://github.com/jenkinsci/docker-workflow-plugin/commit/0b3df67a28cb9993b8585e48c55c4a58246b7cb1).   The bug is that the redirection is causing the exit status to be 1, when it should be 0 in some instances.

To workaround that, I removed the output redirection which restored previous behavior.  See the full commit message for the result of each command.  Here is the result of testing the return value of the command (inside the demo container):

```
root@8a07680574ce:/tmp/files# docker --version
Docker version 1.7.0, build 0baf609
root@8a07680574ce:/tmp/files# docker inspect -f . maven:3.3.9-jdk-8
Error: No such image or container: maven:3.3.9-jdk-8
root@8a07680574ce:/tmp/files# echo $?
1
root@8a07680574ce:/tmp/files# docker inspect -f . maven:3.3.9-jdk-8 >&- 2>&-
root@8a07680574ce:/tmp/files# echo $?
1
# run the job to create the container
root@8a07680574ce:/tmp/files# docker inspect -f . maven:3.3.9-jdk-8
.
root@8a07680574ce:/tmp/files# echo $?
0
root@8a07680574ce:/tmp/files# docker inspect -f . maven:3.3.9-jdk-8 >&- 2>&-
root@8a07680574ce:/tmp/files# echo $?
1
root@8a07680574ce:/tmp/files#
```

This is the error that appeared in the Jenkins job:
```
[workspace] Running shell script
+ docker pull maven:3.3.9-jdk-8
3.3.9-jdk-8: Pulling from library/maven
Digest: sha256:a5283092002ec9739819c9b392203103346dfc163586ff46c30ead13b60cc436
Status: Image is up to date for maven:3.3.9-jdk-8
[Pipeline] withEnv
[Pipeline] {
[Pipeline] withDockerRegistry
[Pipeline] {
[Pipeline] stage (Build)
Entering stage Build
Proceeding
[Pipeline] sh
[workspace] Running shell script
+ docker inspect -f . maven:3.3.9-jdk-8
[Pipeline] sh
[workspace] Running shell script
+ docker pull localhost/maven:3.3.9-jdk-8
Pulling repository localhost/maven
Error: image maven not found
[Pipeline] }
[Pipeline] // withDockerRegistry
[Pipeline] }
[Pipeline] // withEnv
[Pipeline] }
[Pipeline] // node
[Pipeline] End of Pipeline
ERROR: script returned exit code 1
Finished: FAILURE
```

antoniobeyah commented:
@reviewbybees

reviewbybees commented:
This pull request originates from a [CloudBees](https://www.cloudbees.com/) employee. At CloudBees, we require that all pull requests be reviewed by other CloudBees employees before we seek to have the change accepted. If you want to learn more about our process please see [this explanation](https://github.com/reviewbybees/about#about-reviewbybees).

antoniobeyah commented:
@jglick @ndeloof @svanoort @reviewbybees ping

ndeloof commented:
Compared to #25 I can't see user set when `docker exec` is ran, which will cause issues with file permissions on workspace.

antoniobeyah commented:
Good catch, I'll get that added.

antoniobeyah commented:
@ndeloof i added the --user change.  i had some troubles verifying that it is actually doing what I expected (I couldn't directly see the docker exec command) but it seems to pass in the demo container and against some of my existing jenkins jobs that use the plugin. 

I performed an 'id' in the sh step and it was indeed the jenkins user (in my test instance, in the demo container the jenkins user is root).  I tried to figure out how to make the 'docker exec' command it was running print out to the console log like the docker run command but was unsuccessful at figuring out how to do that.

This may be a long-term thing (and probably a mailing list discussion), but I would like to see/determine if the core logic of this plugin and the [Docker Custom Build Environment](https://wiki.jenkins-ci.org/display/JENKINS/CloudBees+Docker+Custom+Build+Environment+Plugin) plugin can be merged so the results are consistent between the two.  I currently use both and the behavior between these two are similar but different enough to make it confusing.

ndeloof commented:
LGTM

@antoniobeyah I would have expected the docker custom build environment plugin to provide `.inside` extension do docker-pipeline, and avoid such code duplication :-\
anyway, imho docker+jenkins future is [docker-slaves-plugin](https://wiki.jenkins-ci.org/display/JENKINS/Docker+Slaves+Plugin).

antoniobeyah commented:
@ndeloof thanks for the review, are there any additional required reviewers or can you merge it?  I don't have merge rights or else I would based on your approval.  I just want to get it off my plate as one less thing to check-in/monitor status of before someone else introduces some conflicting changes to master.  ;)

ndeloof commented:
@antoniobeyah need to wait for plugin maintainer (@jglick)

jglick commented on src/main/java/org/jenkinsci/plugins/docker/workflow/WithContainerStep.java:
```diff
@@ -170,11 +170,11 @@ private static void destroy(String container, Launcher launcher, Node node, EnvV
             ImageAction.add(step.image, run);
             getContext().newBodyInvoker().
-                    withContext(BodyInvoker.mergeLauncherDecorators(getContext().get(LauncherDecorator.class), new Decorator(container, envHost, ws))).
+                    withContext(BodyInvoker.mergeLauncherDecorators(getContext().get(LauncherDecorator.class), new Decorator(container, envHost, ws, dockerClient.whoAmI()))).
```
For safety, run `whoAmI` *before* starting the container, since if it fails we cannot clean up.

jglick commented on src/main/java/org/jenkinsci/plugins/docker/workflow/client/DockerClient.java:
```diff
@@ -87,10 +87,13 @@ public DockerClient(@Nonnull Launcher launcher, @CheckForNull Node node, @CheckF
      * @return The container ID.
      */
-    public String run(@Nonnull EnvVars launchEnv, @Nonnull String image, @CheckForNull String args, @CheckForNull String workdir, @Nonnull Map<String, String> volumes, @Nonnull Collection<String> volumesFromContainers, @Nonnull EnvVars containerEnv, @Nonnull String user, @CheckForNull String... command) throws IOException, InterruptedException {
+    public String run(@Nonnull EnvVars launchEnv, @Nonnull String image, @CheckForNull String args, @CheckForNull String workdir, @Nonnull Map<String, String> volumes, @Nonnull Collection<String> volumesFromContainers, @Nonnull EnvVars containerEnv, @CheckForNull String user, @CheckForNull String... command) throws IOException, InterruptedException {
```
Just delete the argument if it is unused.

jglick commented on src/main/resources/org/jenkinsci/plugins/docker/workflow/Docker.groovy:
```diff
@@ -114,7 +114,7 @@ class Docker implements Serializable {
         public <V> V inside(String args = '', Closure<V> body) {
             docker.node {
-                if (docker.script.sh(script: "docker inspect -f . ${id} >&- 2>&-", returnStatus: true) != 0) {
+                if (docker.script.sh(script: "docker inspect -f . ${id}", returnStatus: true) != 0) {
```
Why?

jglick replied:
Never mind, noted in PR description.

antoniobeyah replied:
This is needed by this PR iirc, it was causing the build to fail when using the demo container.

jglick commented on demo/Dockerfile:
```diff
@@ -26,14 +26,14 @@ FROM jenkinsci/workflow-demo:2.2
-RUN wget -O /usr/bin/docker --no-check-certificate https://get.docker.com/builds/Linux/x86_64/docker-1.4.0
-RUN chmod a+x /usr/bin/docker
+RUN wget --no-verbose -O /usr/bin/docker --no-check-certificate https://get.docker.com/builds/Linux/x86_64/docker-1.7.0 && \
+    chmod a+x /usr/bin/docker
```
Are these related somehow?

jglick replied:
I guess only in that the client version is being upgraded to 1.7.0.

antoniobeyah replied:
This change was for three reasons:
1) Upgrade the docker client to v1.7.0 (the important change)
2) To get rid of the scrolling wall of text that occurred due to the output expecting to be interactive.  It was printing a statusbar to the log on a given interval
3) Prevent docker from caching unnecessary additional layers.  The Dockerfile best practices recommend concatenating your run commands to prevent caching of some layers.  I stopped short of doing it to the entire file and settled for the two related items.  See: [Dockerfile best practices](https://docs.docker.com/engine/userguide/eng-image/dockerfile_best-practices/)

jglick commented:
So is this purporting to address [JENKINS-34289](https://issues.jenkins-ci.org/browse/JENKINS-34289) or not? And is there a self-contained demo case that can be used to prove that this change fixes some concrete bug? What I am missing from all this is a clear description of what problem is being solved.

jglick commented on src/main/java/org/jenkinsci/plugins/docker/workflow/WithContainerStep.java:
```diff
@@ -200,17 +200,24 @@ private static FilePath tempDir(FilePath ws) {
-                    List<String> prefix = new ArrayList<String>(Arrays.asList("docker", "exec", container, "env"));
+                    List<String> prefix;
+                    if (userId == null) {
+                        prefix = new ArrayList<String>(Arrays.asList("docker", "exec", container, "env"));
```
Good, this is necessary for compatibility with builds running before the upgrade.

jglick commented on src/test/java/org/jenkinsci/plugins/docker/workflow/client/DockerClientTest.java:
```diff
@@ -76,6 +76,26 @@ public void test_run() throws IOException, InterruptedException {
+        Assert.assertTrue(containerRecord.getContainerName().length() > 0);
+        Assert.assertTrue(containerRecord.getHost().length() > 0);
+        Assert.assertTrue(containerRecord.getCreated() > 1000000000000L);
+        Assert.assertEquals(Collections.<String>emptyList(), dockerClient.getVolumes(launchEnv, containerId));
```
Not really testing anything interesting here; maybe better to just delete this, especially if the `user` argument is deleted.

What I would like to see in a test (if it is possible) is verification in `WithContainerStepTest` that you are fixing some bug: a script which fails before the switch of the user ID from `run` to `exec` but which passes afterwards.

antoniobeyah replied:
The original intent was to test that not passing in a user would not define a user, but the ContainerRecord doesn't provide a user field so I settled for just ensuring that when passing a null value into the user field you still received a valid ContainerRecord.  I can remove this if you want.

jglick commented:
> I found a bug in master with docker client 1.7.0, the bug was introduced in #54. The bug is that the redirection is causing the exit status to be 1, when it should be 0 in some instances.
> 
> To workaround that, I removed the output redirection which restored previous behavior.

This sounds like a separate issue that would better be fixed in its own pull request. (If basic testing of _this_ PR requires that fix, just `merge` in that branch.)

jglick commented:
> I performed an 'id' in the sh step and it was indeed the jenkins user (in my test instance, in the demo container the jenkins user is root).

This is an example of something you have apparently tested interactively but which is not reflected in an automated test, or even written manual test instructions.

jglick commented:
> I would like to see/determine if the core logic of this plugin and the Docker Custom Build Environment plugin can be merged

Given how fundamentally different Pipeline asynchronous (durable) process execution is from freestyle, I do not think this is practical.

ndeloof commented:
@jglick both rely on `docker exec` a very comparable way, I don't see any reason they don't share a common code base.

antoniobeyah commented:
@jglick  JENKINS-34289 was marked as resolved so I ignored it for the purposes of this PR.  Point taken on the clear explanation on why this is needed, I will try to get a test case written up that shows the failure.  

I have a limited amount of time I can spend on this- we are running on a forked version at Target that has the changes in it (not ideal).  Are you able to provide a bulleted list of items that you require for a merge?  So far I have gathered:

1) Move the master bug fix to a seperate pr
2) Provide a test showing the thing this PR is intended to fix
3) (breaks backwards compability) Remove the 'user' parameter from  DockerClient#run
4) Move dockerClient.whoAmI() to occur before the invocation

Are there any additional changes you would require?

jglick commented:
> Move the master bug fix to a seperate pr

Well, desirable practice generally, so that the obvious bug fix can get to users while we are discussing the main part of this PR. Can be skipped if you do not have time.

> Provide a test showing the thing this PR is intended to fix

Yes, manual test steps if it is impractical to write an automated test (for example because you need to be using a specific user ID on the host).

> breaks backwards compability

It is fine, `DockerClient` is really just an internal utility for this plugin.

> Move dockerClient.whoAmI() to occur before the invocation

Right, safer.

> Are there any additional changes you would require?

No, the test—or just some demonstration that this is fixing a particular bug I can see for myself—is the main thing.

dweomer commented:
I do not understand why removing the user constraints put on a container for `docker.inside` is under consideration. The `docker.inside` mechanism is for injecting a workspace into a container so that it may act upon it. If a container cannot run as the specified user then it's image is simply not fit for this purpose. After injecting the workspace into the container to mutate it I expect that follow-on processes will be able to further mutate it as will. **_Choosing to forgo the specification of the user that the container should run as breaks this very fundamental expectation._**

@carlossg pointed me here from carlossg/docker-maven#16, and yes this would fix the problem but at the cost of a good deal more downstream breakage. What this PR purports to fix, that being #25, seems of dubious value to me. I think it would make more sense to leave the default behavior of specifying the user as-is unless the image specified for `docker.inside` is specially packaged, say with labels, to support advanced functionality. This could be as simple as having a label such as `io.jenkins.pipeline.docker.inside.user=root` on the container. This would not, of course, cover the desired contract that the resulting workspace can still be written to by the invoking user but given that `docker.inside` is already docker-aware I imagine it could perform some nifty post processing (or possibly tricksy user namespace remapping) to have the output workspace owned and writable by the invoking user, as would be expected.

antoniobeyah commented:
@dweomer The constraint isn't being removed, it is being moved to the exec command.  This will allow the container to start up as expected, without unintended side effects.

I've been slow on getting this pr updated with the requests, but this is our use case:

We use docker images for CI testing of RPM builds.  The RPMs happen to use systemd.  That means entrypoint is 'init' which must be run as root, otherwise the container blows up.  All subsequent commands are executed as the jenkins user, and that is supported by 'docker exec --user <user>'.  The container behaves very different when it is started with 'docker run --user' vs 'docker exec --user'.  You are unable to do certain things when --user is used on the 'docker run', such as using sudo to install any required packages using yum, or to get the container to start at all.

This behavior is already supported by other plugins, such as [Docker Custom Build Environment](https://github.com/jenkinsci/docker-custom-build-environment-plugin) plugin.  We are running a forked version of the docker-pipeline plugin without the docker run --user constraint and it is working great.

This PR only allows for the possibility of using additional containers without the need of custom tailoring them to be used within Jenkins.  I would also be for using a different mechanism for determining the user to run .inside as, but as it stands it works 'well enough' with the changes in this PR.

@dweomer Will this PR break your environment somehow?  If so I may be able to add in some checks to make that less likely, just need some specific details on your concerns.

@jglick As far as the automated test- I don't think I will be able to write one due to the reasons you provided (the user in the host machine is root, my test requires it to be a non-root user).  Can you provide more details on how specifically I can provide manual verification steps?

jglick commented:
> Can you provide more details on how specifically I can provide manual verification steps?

Well, something I can run locally (perhaps inside VirtualBox if host permissions matter) that would let me understand the problem for myself and evaluate how the proposed fix works. Minimally, a demo Pipeline script relying only on images on Docker Hub (can use `.build` as needed to create derivative images). If I can reproduce it manually, I at least have a shot at adding an automated test as well.

jglick commented:
> Will this PR break your environment somehow? If so I may be able to add in some checks to make that less likely

If there is any concern about this, I would rather make the new behavior opt-in.

dweomer commented:
@antoniobeyah wrote:
> Will this PR break your environment somehow? If so I may be able to add in some checks to make that less likely, just need some specific details on your concerns.

In my ire related to carlossg/docker-maven#16, I had forgotten about the run/exec distinction. This should not break environments but I feel that you should ensure that by implementing a test case such as:
- pull down well known or sample project
- build it via `docker(img).inside` as appropriate (for each of golang, maven, and possibly node/npm?)
- wipe out the entire workspace without encountering write permission failures

dweomer commented:
@antoniobeyah @jglick:
My other qualm with this is that I do not want my build containers to run as root unless I explicitly tell them to via code (as that implies pipeline code which is something that can be audited). Currently most official images for build systems such as `golang` and `maven` run as root by default (which is correct because these are very low level base images) and this is currently overriden by `docker.inside` by default to match the user:group of the process that invokes the in-container build.

In short, defaulting to the image's `USER` introduces a potential security hole as a side-effect which is not acceptable. Opt-in semantics would address this issue.

dweomer commented:
@antoniobeyah after re-reading your use-case I don't see that you need make any code changes at all, you just need to specify the user a la `docker.image('rpm-with-systemd-tester').inside('--user root:root')`. The below example works on both a Jenkins 1.x LTS as well as 2.x LTS instance:
```
timestamps { -> node {
    stage ('DEFAULT') {
        docker.image('maven:3.2').inside('--group-add 999') {
            sh 'id'
        }
    }
    stage ('ROOT') {
        docker.image('maven:3.2').inside('--user root:root') {
            sh 'id'
        }
    }
}}
```
with output that looks like:
```
Started by user admin
[Pipeline] timestamps
[Pipeline] {
[Pipeline] node
00:38:05 Running on master in /var/jenkins_home/workspace/maven-3-point-2
[Pipeline] {
[Pipeline] stage
[Pipeline] { (DEFAULT)
[Pipeline] sh
00:38:05 [maven-3-point-2] Running shell script
00:38:05 + docker inspect -f . maven:3.2
00:38:05 .
[Pipeline] withDockerContainer
00:38:05 $ docker run -t -d -u 1000:1000 --group-add 999 -w /var/jenkins_home/workspace/maven-3-point-2 --volumes-from dc8e94e8e3eaa302a91ebb2458fd093ffc9287e3992ee37cbe2e5392361b5532 -e ******** -e ******** -e ******** -e ******** -e ******** -e ******** -e ******** -e ******** -e ******** -e ******** -e ******** -e ******** --entrypoint cat maven:3.2
[Pipeline] {
[Pipeline] sh
00:38:06 [maven-3-point-2] Running shell script
00:38:07 + id
00:38:07 uid=1000 gid=1000 groups=1000,999
[Pipeline] }
00:38:07 $ docker stop 2e1c73c55e2d4b73461f40ba14f4df7fdeece2ecb0075f1ce546cbc6e261ad28
00:38:17 $ docker rm -f 2e1c73c55e2d4b73461f40ba14f4df7fdeece2ecb0075f1ce546cbc6e261ad28
[Pipeline] // withDockerContainer
[Pipeline] }
[Pipeline] // stage
[Pipeline] stage
[Pipeline] { (ROOT)
[Pipeline] sh
00:38:17 [maven-3-point-2] Running shell script
00:38:18 + docker inspect -f . maven:3.2
00:38:18 .
[Pipeline] withDockerContainer
00:38:18 $ docker run -t -d -u 1000:1000 --user root:root -w /var/jenkins_home/workspace/maven-3-point-2 --volumes-from dc8e94e8e3eaa302a91ebb2458fd093ffc9287e3992ee37cbe2e5392361b5532 -e ******** -e ******** -e ******** -e ******** -e ******** -e ******** -e ******** -e ******** -e ******** -e ******** -e ******** -e ******** --entrypoint cat maven:3.2
[Pipeline] {
[Pipeline] sh
00:38:19 [maven-3-point-2] Running shell script
00:38:19 + id
00:38:19 uid=0(root) gid=0(root) groups=0(root)
[Pipeline] }
00:38:19 $ docker stop 34e4441d70587203e1b0ad047e4c6363b33588d36f9e7535eb00200255ecf54c
00:38:30 $ docker rm -f 34e4441d70587203e1b0ad047e4c6363b33588d36f9e7535eb00200255ecf54c
[Pipeline] // withDockerContainer
[Pipeline] }
[Pipeline] // stage
[Pipeline] }
[Pipeline] // node
[Pipeline] }
[Pipeline] // timestamps
[Pipeline] End of Pipeline
Finished: SUCCESS
```

dweomer commented:
The trick above is that `docker run` seems to be forgiving of multiple `--user` specifications and the last one in wins. I don't run RPM based images and so if you are using RedHat's docker fork this may not be supported but it seems worth a try.

jglick commented:
> you just need to specify the user a la `docker.image('rpm-with-systemd-tester').inside('--user root:root')`

Interesting, though it would be nicer to have this be an advertised option.

[JENKINS-38438](https://issues.jenkins-ci.org/browse/JENKINS-38438) was recently filed, which sounds like it is on the same topic, yet switching the `--user` from `docker-run` to `docker-exec` does not make any difference in that case: either way, the command fails.

docwhat commented:
Some containers come with pre set users, etc. The automatic adding of volumes and users can be disruptive and hard for users to troubleshoot. 

Can we split a plain docker environment out from the mount everything as a workspace version?

Or should everyone fall back to `sh 'docker...'`?

antoniobeyah commented:
@jglick I have rebased the branch and added in test cases to demonstrate the use case for this pr.  I also moved the dockerClient.whoAmI() to occur before the container run.  I didn't add any guards to it because whoAmI will throw an exception on error.  I ran the tests, but they fail and they are all appear to be setup as 'skip' when run through maven so I am assuming there is some trick to getting them to pass.  The error I get is below.  The latest entrypoint changes on master also cause an issue. The start command (entrypoint) for my container must be '/usr/sbin/init', but that is currently overridden by 'cat' due to entrypoint positioning.  I am looking to see if there is anything I can do to prevent that.  I will likely move the default stuff to happen before user defined stuff.

The test cases don't exactly illustrate whats happening below, I will attempt to get them closer to the manual testing steps I performed below.

The differences:
1) --entrypoint is 'cat' in the tests due to latest commits in master
2) The tests relies on the fact that they are currently executing as root (dockerClient.whoAmI() will return 0:0), so the tests should work but they are not portable if the user running the tests is not root

Manual steps:
### Successful Case (no --user on docker run)
```
-> docker run -t -d -v /sys/fs/cgroup:/sys/fs/cgroup --privileged --entrypoint /usr/sbin/init centos/systemd
1520faa10bcc31295a9ee08b8c29bf71b13f246a17296619b6a38bcfcfc3b340
ruby-2.3.1 ~ $
-> docker exec -it 1520faa10bcc31295a9ee08b8c29bf71b13f246a17296619b6a38bcfcfc3b340 bash
[root@1520faa10bcc /]# systemctl status
● 1520faa10bcc
    State: running
     Jobs: 0 queued
   Failed: 0 units
    Since: Sat 2016-09-24 18:43:08 UTC; 13s ago
   CGroup: /
           ├─ 1 /usr/sbin/init
           ├─17 bash
           ├─30 systemctl status
           ├─31 systemctl status
           └─system.slice
             └─systemd-journald.service
               └─16 /usr/lib/systemd/systemd-journald
[root@1520faa10bcc /]#
```

### Failing Case (arbitrary --user on docker run)
```
-> docker run -t -d -v /sys/fs/cgroup:/sys/fs/cgroup --privileged --entrypoint /usr/sbin/init --user 1234:1234 centos/systemd
68c891f1058439740d6fb674adf66f1ee8a46dff185dde11069a90ca93c1cb4c
ruby-2.3.1 ~ $
-> docker exec -it --user 0 68c891f1058439740d6fb674adf66f1ee8a46dff185dde11069a90ca93c1cb4c bash
[root@68c891f10584 /]# systemctl status
Failed to get D-Bus connection: Operation not permitted
[root@68c891f10584 /]#
```


```
=== Starting withExecAsUser(org.jenkinsci.plugins.docker.workflow.WithContainerStepTest)
[prj #1] Started
Sep 24, 2016 1:32:45 PM hudson.ExtensionFinder$GuiceFinder$FaultTolerantScope$1 error
WARNING: Failed to instantiate optional component org.jenkinsci.plugins.workflow.steps.scm.SubversionStep$DescriptorImpl; skipping
[prj #1] [Pipeline] node
[prj #1] Running on master in /var/folders/wl/kkhnc8fx0zq1lj1_kqd1bszc0000gp/T/junit4999253668053292568/junit6145851350636653994/workspace/prj
[prj #1] [Pipeline] {
[prj #1] [Pipeline] withDockerContainer
[prj #1] $ docker run -t -d -v /sys/fs/cgroup:/sys/fs/cgroup:ro --privileged -w /var/folders/wl/kkhnc8fx0zq1lj1_kqd1bszc0000gp/T/junit4999253668053292568/junit6145851350636653994/workspace/prj -v /var/folders/wl/kkhnc8fx0zq1lj1_kqd1bszc0000gp/T/junit4999253668053292568/junit6145851350636653994/workspace/prj:/var/folders/wl/kkhnc8fx0zq1lj1_kqd1bszc0000gp/T/junit4999253668053292568/junit6145851350636653994/workspace/prj:rw -v /var/folders/wl/kkhnc8fx0zq1lj1_kqd1bszc0000gp/T/junit4999253668053292568/junit6145851350636653994/workspace/prj@tmp:/var/folders/wl/kkhnc8fx0zq1lj1_kqd1bszc0000gp/T/junit4999253668053292568/junit6145851350636653994/workspace/prj@tmp:rw -e ******** -e ******** -e ******** -e ******** -e ******** -e ******** -e ******** -e ******** -e ******** -e ******** -e ******** -e ******** -e ******** -e ******** -e ******** -e ******** --entrypoint cat centos/systemd
[prj #1] [Pipeline] {
[prj #1] [Pipeline] sh
[prj #1] [prj] Running shell script
[prj #1] sh: /var/folders/wl/kkhnc8fx0zq1lj1_kqd1bszc0000gp/T/junit4999253668053292568/junit6145851350636653994/workspace/prj@tmp/durable-18bae9d0/pid: No such file or directory
[prj #1] sh: /var/folders/wl/kkhnc8fx0zq1lj1_kqd1bszc0000gp/T/junit4999253668053292568/junit6145851350636653994/workspace/prj@tmp/durable-18bae9d0/jenkins-log.txt: No such file or directory
[prj #1] sh: /var/folders/wl/kkhnc8fx0zq1lj1_kqd1bszc0000gp/T/junit4999253668053292568/junit6145851350636653994/workspace/prj@tmp/durable-18bae9d0/jenkins-result.txt: No such file or directory
[prj #1] [Pipeline] }
[prj #1] $ docker stop 323ff2f8568abe5aacea287c263290428a37c90abacbfa60862d6ff82c419f5e
[prj #1] $ docker rm -f 323ff2f8568abe5aacea287c263290428a37c90abacbfa60862d6ff82c419f5e
[prj #1] [Pipeline] // withDockerContainer
Sep 24, 2016 1:33:13 PM org.jenkinsci.plugins.workflow.job.WorkflowRun finish
INFO: prj #1 completed: FAILURE
[prj #1] [Pipeline] }
[prj #1] [Pipeline] // node
[prj #1] [Pipeline] End of Pipeline
[prj #1] ERROR: script returned exit code -2
[prj #1] Finished: FAILURE
```

@dweomer That sounds like it may fit your use case, but not mine.  I want the containers to start up as they were built, not with an arbitrary override.  I need the distinction of run vs exec due to us not running our jenkins agents as root (we have a jenkins user).

@docwhat I agree with you, but I think it may be best for you to open your own issue for that discussion.  I would like to prevent this PR (which I feel addresses an immediate need of myself and the others who have commented) from getting bogged down in a design discussion.  It would certainly stop the assumptions of the plugin from interfering with the container.

dweomer commented:
@antoniobeyah wrote:
> @dweomer That sounds like it may fit your use case, but not mine. I want the containers to start up as they were built, not with an arbitrary override. I need the distinction of run vs exec due to us not running our jenkins agents as root (we have a jenkins user).

Specifically:
>  I want the containers to start up as they were built, not with an arbitrary override

This tells me that vanilla `docker.inside` isn't a good fit for your use case and that `docker.run` combined with explicit `sh "docker exec -it $cid some-useful-command"` is a better fit.

Also:
> That sounds like it may fit your use case, but not mine.

I believe that if it weren't for the advent of the `--entrypoint cat` (a recent-ish addition I think) from `docker.inside` that it would work for you without issue, as my example above shows. This again speaks to vanilla `docker.inside` being a poor fit for your use case. Have you tried something like:
```
docker.image('centos/systemd').inside('--privileged --user root:root --volume /sys/fs/cgroup:/sys/fs/cgroup') {
    sh '/usr/sbin/init &'
    sh '# useful work goes here'
}
```
Well, wait. Of course that will not work because your image's entry-point must needs be PID 1. Sounds like `docker.inside` simply needs to support overrides for `--user` and `--entrypoint` and then with a little work and no custom plugin your scripts will work AND my scripts will continue to work with the already established, herein unmodified, implicit build-time contract. What I am proposing might look like:
```
docker.image('centos/systemd').user('root').entrypoint('/usr/sbin/init').inside('--privileged --volume /sys/fs/cgroup:/sys/fs/cgroup') {
    sh '# useful work goes here'
}
```
---
I appreciate that you do not wish your PR to get bogged down in design discussion but your changes, without an opt-in (aka moot the arbitrary override with a specified one), will result in just about any `Jenkinsfile` that expects to be able to write to it's workspace after executing `docker.inside` to fail horribly. If, after review, @jglick and the other CloudBees folks decide to break such an important, albeit **_implicit_**, contract I will be sorely disappointed but at least it will have gone through said review.

dweomer commented:
@jglick wrote:
> Interesting, though it would be nicer to have this be an advertised option.
> 
> [JENKINS-38438](https://issues.jenkins-ci.org/browse/JENKINS-38438) was recently filed, which sounds like it is on the same topic, yet switching the --user from docker-run to docker-exec does not make any difference in that case: either way, the command fails.

[JENKINS-38438](https://issues.jenkins-ci.org/browse/JENKINS-38438) is a result of tools like `ssh` ignoring the value of `$HOME` and instead forcing de-reference of the user's home directory from `/etc/passwd` which fails miserably if there is no entry for the runtime UID. One assumes there is an underlying security reason for this behavior from `ssh`. This hints at missing support, in `docker.inside` and similar, for automagic setup of an "arbitrary" user, in this default case the user invoking the "build".

dweomer commented:
@antoniobeyah, I wrote:
> I appreciate that you do not wish your PR to get bogged down in design discussion **but your changes, without an opt-in (aka moot the arbitrary override with a specified one), will result in just about any `Jenkinsfile` that expects to be able to write to it's workspace after executing `docker.inside` to fail horribly**. If, after review, @jglick and the other CloudBees folks decide to break such an important, albeit **_implicit_**, contract I will be sorely disappointed but at least it will have gone through said review.

... and after reviewing your code changes, this part of what I wrote is simply wrong (bolded above):
> but your changes, without an opt-in (aka moot the arbitrary override with a specified one), will result in just about any `Jenkinsfile` that expects to be able to write to it's workspace after executing `docker.inside` to fail horribly

As you pointed out, however, you still need support for a custom `--entrypoint` which seems to fit well with supporting a custom `--user`.

jglick commented:
> Can we split a plain docker environment out from the mount everything as a workspace version?

That would be `Image.run/withRun`.

jglick commented:
There is a lot of information here which will clearly take me some time to digest, experiment with, etc. Right now I cannot spend time on it. Whether a change to default behavior, new option, or simply fuller documentation is most appropriate here is TBD.

docwhat commented:
@jglick: I tried putting some of my thoughts about the docker interfaces into [JENKINS-38620](https://issues.jenkins-ci.org/browse/JENKINS-38620). I hope they help.

jglick commented:
Does it make sense to fold this into #78, so that switching to a new command disables both `ENTRYPOINT` override and user override? Is it basically the same use case, as is hinted at by

> The latest entrypoint changes on master also cause an issue.

?

rymndhng commented:
Any update on this issue -- I've been working around this by creating a user with the UID that jenkins appears to be using (which is a PITA)

chancez commented:
@rymndhng how are you doing that, my image isn't running as root, so I can't figure out how to create the user properly.

rymndhng commented:
@chancez I add something like this to my dockerfile

``` Dockerfile
RUN useradd -u 1001 -m build-user
USER build-user
```

Replace `1001` with the UID that you're having issues with. You can also name `build-user` whoever you want.

chancez commented:
Ah, I see. I was thinking you were doing it in your Jenkinsfile.

rymndhng commented:
@chancez yeah, I've had no success fixing it from the Jenkinsfile 😢

davidmnoriega commented:
Blundered into this today. @dweomer is right, passing `-u root` to `inside|run|withRun()` in my jenkinsfile as a workaround works because it does indeed look like the last `-u` in the `docker run` command wins. Now, how long that continues to work...

jglick commented:
Generally if your `Dockerfile` includes a `USER` command, you should not be using `Image.inside`. It is just not designed for such cases.

docwhat commented:
> Generally if your `Dockerfile` includes a `USER` command, you should not be using `Image.inside`. It is just not designed for such cases.

Can we use `docker inspect` and issue a warning for that case?

zasdfgbnm commented:
@jglick it might be better to add `-v /etc/group:/etc/group:ro -v /etc/passwd:/etc/passwd:ro` if the final decision is to keep the `--user`

joshuaspence commented:
Is this PR going anywhere?

michaelneale commented:
doing a sweep of older PRs and it looks like this one is no longer live, and there are sufficient work arounds - can we close this?

ndeloof commented:
@michaelneale I still think we need this one, especially if we restore `ENTRYPOINT` support then this one will expect to run with userID as designed in docker image (let's say, to start background services)

michaelneale commented:
@ndeloof @antoniobeyah want to bring this one up to date then?

adamvoss commented:
All the discussion here has been with the Scripted Pipeline.  Declarative pipeline is also affected.

For example, to make the Python image work, you need to add `args '--user root:root'`

Example:
```groovy
agent {
  docker {
    image 'python:2'
    args '--user root:root'
  }
}
```

I find this rather unexpected and confusing behavior when commands will work directly if using the container on CLI but don't under Jenkins and it has taken me quite a while (~a month) to what is going on, find this issue, and know how to make all builds consistently work.

My gut reaction is to agree that Jenkins should use the user from the container unless specified otherwise.  This would make the _naive_ declarative Jenkinsfile just work with images pulled from Dockerhub and wanting different behavior could still specify so.  As such I  would like to see this go forward, but based on the discussion above (and the `wontfix` tag), I'm not sure if there is support for the change of behavior.

bertjwregeer commented:
The problem of the Jenkins uid/gid outside the container not matching anything in `/etc/passwd` inside the container is a problem I've been running into, and the decision made in https://issues.jenkins-ci.org/browse/JENKINS-41050 to do the checkout outside means that files created inside the container when running as "root" inside the container (as @adamvoss points out for the Python images) can no longer be cleaned up by Jenkins, so git clean fails, and fails hard (see https://issues.jenkins-ci.org/browse/JENKINS-24440).

fessmage commented:
By my opinion, default behaviour of running docker images from jenkins pipeline must be the same, what you expect and what you get, when running same docker images on your local workspace. 
At least current situation must be documented and contain examples - because for working around with what's going on, why you can't reproduce workflow in pipeline - take times!

Atrifex commented:
What is the ETA for the resolution of this issue?

qbaze commented:
Hey guys.

Why is this one 'wontfix'? Is it a desired behaviour? What's the purpose if yes?
Sorry if I missed something.

Best,
Piotr

dmolesUC commented:
@michaelneale what are the "sufficient workarounds"?

jglick commented:
Should be closed, risk of regression is too high. The workaround is to simply not use `withDockerContainer` in even a remotely subtle case. If it works the first time you try it, great. If not, use `sh 'docker …'` instead and forget about this misconceived feature.

metametadata commented:
This is my current code, not sure how it can be replaced with `sh 'docker …'`:

```
image.inside(...

            // HACK: setting root user is needed because otherwise the current user id from the host will be used. See:
            //   https://issues.jenkins-ci.org/browse/JENKINS-38438
            //   https://github.com/jenkinsci/docker-workflow-plugin/pull/57
            // Without it something as simple as 'mkdir foo' unexpectedly fails because of permission denied.
            "-u root " +

            ...

            ) {
        try {
            stage('...') {
                ...
            }

            stage('...') {
                ...
            }

            body()
        }
        finally {
            // HACK: fix permissions because Docker creates root-owned files which cannot be deleted
            // by Jenkins user when it tries to clean the workspace.
            // https://issues.jenkins-ci.org/browse/JENKINS-24440#comment-335677
            // https://issues.jenkins-ci.org/browse/JENKINS-46686
            sh('chmod -R 777 .')
        }
```

bertjwregeer commented:
Using `sh docker ...` instead (i.e. invoking docker oneself) doesn't solve the problem of still needing access to the checked out source code.

Unless you are suggesting we check out the source code inside the docker container, because doing a bind mount (which is what `image.inside` is effectively doing) would still run into issues with file permissions.

jglick commented:
There is no universal advice, but if you need access to files from `checkout scm` or the like, it is typically most straightforward to reconceive the build process as a `docker build` command with a `Dockerfile` using `COPY` to pick up files from a relative location. This also avoids accidentally polluting the workspace with files owned by other UIDs; on the flip side, if you need some build products in the workspace in order to run Jenkins publishers, you will need to use `docker cp` as shown in #105.

zerkms commented:
> if you need access to files from checkout scm or the like, it is typically most straightforward to reconceive the build process as a docker build command with a Dockerfile using COPY to pick up files from a relative location

This is a horrible workaround: given in other cases workspace is shared automatically - doing something manually anywhere is a recipe for a disaster.

bertjwregeer commented:
Copying large source trees into a `docker build` only to then use that docker image is also expensive.

Yet, in most cases I'll need the source code checked out to be able to use the included `Dockerfile` to now build the docker image so that I can build the source code and pull artifacts out.

arowland-informatics commented:
This bug still affects declarative pipelines and essentially makes using a docker agent not possible. What would be the workaround if you are not using scripted pipelines?

RedDwarf69 commented:
There is https://issues.jenkins.io/browse/JENKINS-63233. In a rootless podman container you are root insider the container, but jenkins outside. So the workspace file permission issue doesn't exist.

I guess the same applies to https://docs.docker.com/engine/security/rootless/. But you may have less control over the user the Docker daemon is running as while with podman, being daemon-less, it just runs as the user calling it (i.e. jenkins).

iliis commented:
I'm using simple declarative pipelines like this

```groovy
pipeline {
    agent {
        dockerfile true
    }

    stages {
		...
	}
}
```

or like this:

```groovy
pipeline {
    agent {
        docker {
            image "some-predefined-docker-image"
        }
    }

    stages {
		...
	}
}
```

as described in https://www.jenkins.io/doc/book/pipeline/docker/

However, due to this issue I need to add the following to all my `Dockerfiles`:

```Dockerfile
RUN groupadd --gid 1000 jenkins && \
    useradd --create-home --uid 1000 --gid 1000 --groups sudo jenkins && \
    echo "jenkins ALL=(ALL) NOPASSWD: ALL" >> /etc/sudoers
```

Which breaks as soon as I run Jenkins on a machine where the user `jenkins` doesn't have ID 1000, for example when installing it somewhere where there's already another default user that uses this ID.

konserw commented:
Hi! Is there any progress with this MR? I've just been hit by issue with unknow uid in container when run via jenkins. I would also appreciate if jenkins documentation would mention that jenkins is overriding uid in my containers. I'll try out workarounf from previous post for now

diorcety commented:
My PR #160 try to resolve that issue.
I will try to summerize the issue.

The plugins do the following:
Run the image with the following command : docker run -u jenkins cat
Run the commands with the following command: docker exec CMD

The docker behaviour is that, the commands run with exec, without any override will be executed with the same user specifed at docker run: https://stackoverflow.com/questions/35734474/connect-to-docker-container-as-user-other-than-root

In order to use an other user, in the pipeline, he can add another "-u user:user" parameter in order to override the default plugin behaviour. The issue is in that case all the next exec commands will be also executed with that user, which for most the case is not wanted. 
One solution should only be to also put the -u jenkins for every docker exec command decorated. Without any override of user in the run arguments the behaviour will be exactly the same. But this will allow to enter in the image using another user (for exemple root) doing some action in order to create the same user that the host one (like the entrypoint of dockcross https://github.com/dockcross/dockcross/blob/master/imagefiles/entrypoint.sh
) and run all the commands with an equivalent jenkins in the docker. All the files created will be owned by the jenkins user, in the docker or on the host. 
If the current behaviour is needed at the end, that can be done optionnally following a parameter.

jacobnollette commented:
Before I was shown this issue, I had a workaround documented here -
https://stackoverflow.com/questions/74646034/jenkins-docker-pipeline-not-passing-exit-codes

Wondering if there's a better solution to my hacky approach or if there's a better solution coming from this PR? Any updates since last year?

ketozhang commented:
For the love of xyz, document this behavior...if you're not going to merge this.

Do the example code in this documentation even work as-is?
https://www.jenkins.io/doc/book/pipeline/docker/#dockerfile

jglick commented:
> Do the example code in this documentation even work as-is (without setting the `-u` flag to something else)?

Yes.

reitzig commented:
Every year or so, I get hit by this.

Instead of adding creative invective, I'd like to add this factoid: It's not (only) about `root` vs `1000`. Yes, generally speaking, we shouldn't run stuff as `root`. Agreed. Hower, just because the "default" non-root user _happens_ to have UID `1000` for many Linux distros, that choice is by no means universal. One should not rely on that UID to exist, or be the correct one to use, ever.

Good images _do_ set a non-root- user as default user. Have been for years. Jenkins resp. this plugin are not responsible for enforcing that, let alone making an arbitrary decision about what that UUID should be.

FWIW, my case is even worse: there's config baked into the image's default user's home directory; Jenkins tries to use another UID. Even if I hack a matching user into the image, all that config doesn't apply! (Ironically, my image _has_ user `1000` set up, but for some reason, the admins made that Jenkins instance run containers as `1001:1002` 🤪)

> > Do the example code in this documentation even work as-is (without setting the -u flag to something else)?
> Yes.

By happy accident! The _minute_ `node` changes something relevant in their setup, they break.
Need some popular examples? `alpine`; `nginx` and, yes, also `nginxinc/nginx-unprivileged`; `python`; the list goes on. None of these have a user with UID `1000`.

I agree that the current behaviour of the plugin is bad, and I _strongly_ agree that the docs need to reflect the behaviour. 

And now I'm off to try and test the images I build for OpenShift _without_ wrangling them so they conform with an overly opinionated plugin.
